doc/2/api/controllers/document/upsert/index.md

```diff
@@ -0,0 +1,107 @@
+---
+code: true
+type: page
+title: upsert
+---
+
+# upsert
+
+Applies partial changes to a document. If the document doesn't already exist, a new document is created.
+
+---
+
+## Query Syntax
+
+### HTTP
+
+```http
+URL: http://kuzzle:7512/<index>/<collection>/<_id>/_upsert[?refresh=wait_for][&retryOnConflict=<int>][&source]
+Method: PUT
+Body:
+```
+
+```js
+{
+  changes: {
+    // document partial changes
+  },
+  default: {
```

Comment:
i'm personnaly not found of `default` property. IMO whole document to create if needed should be inside. Is there a specific use case for that ?

Reply:
If we want to create the whole document, there is already a `document:createOrReplace` API action available (or `document:replace` if we know that the document already exists).

The use case for an upsert API action has been explained in the PR description. 

As for the `default` object, the reasoning behind this option is this: what happens if we need to update non-required fields, but the document needs to have required values because of business rules or because of strict mappings?
We might need to also provide a few required fields along with the document. And since we don't know if the document exists or not, those values will probably be default ones, meaning that we DON'T want to update them if these fields already exist.

Thus I divided the request payload into two parts: the fields we want to update and, optionally, a set of default fields values that will only be used if the document needs to be created.

Does that answer your question?

Reply:
It does indeed.

> As for the default object, the reasoning behind this option is this: what happens if we need to update non-required fields, but the document needs to have required values because of business rules or because of strict mappings?

I was mitigated because this mode of operation differs from upstert methods in general. (eg: Mongo, etc.), but it is true that the properties that are required justify the choice made.

Reply:
That's probably because traditional upserts work with complete documents (as with our own "createOrReplace")

I don't know about mongo, but that's the case with traditional DBs :wink: 

Our update action takes only partial changes to apply to documents, and until now, we didn't have an "updateOrReplace" action because there weren't proper use cases for it. That kind of action is a bit rare and counterintuitive at first, and it's only because one of our clients needs it that I added it to our API.